server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -205,7 +205,7 @@ public CachingClusteredClient(
 
       // Pull cached segments from cache and remove from set of segments to query
       final Map<Cache.NamedKey, byte[]> cachedValues;
-      if (useCache) {
+      if (useCache && cacheKeys.size() <= cacheConfig.getCacheBulkMergeLimit()) {
```

Comment:
Yeah, I didn't worry about cache keys because they are relatively cheap to calculate.